perfkitbenchmarker/providers/azure/azure_disk.py

```diff
@@ -104,80 +109,79 @@ def __init__(self, disk_spec, vm_name, machine_type):
+      _, _, retcode = vm_util.IssueCommand(
+          [azure.AZURE_PATH, 'vm', 'disk', 'attach-new',
+           '--host-caching', self.host_caching,
+           self.vm_name, str(self.disk_size)] +
+          self.resource_group.args)
+
+      if retcode:
+        raise errors.Resource.RetryableCreationError(
+            'Error creating Azure disk.')
+
       self.created = True
 
+    @vm_util.Retry()
+    def GetName():
+      disk = self._GetDiskJSON()
+      self.name = disk['name']
+    GetName()
+
   def _Delete(self):
     """Deletes the disk."""
-    delete_cmd = [AZURE_PATH,
-                  'vm',
-                  'disk',
-                  'delete',
-                  '--blob-delete',
-                  self.name]
-    logging.info('Deleting disk %s. This may fail while the associated VM '
-                 'is deleted, but will be retried.', self.name)
-    vm_util.IssueCommand(delete_cmd)
-
-  def _Exists(self):
-    """Returns true if the disk exists."""
-    if self.name is None and self.created:
-      return True
-    elif self.name is None:
-      return False
-    show_cmd = [AZURE_PATH,
-                'vm',
-                'disk',
-                'show',
-                '--json',
-                self.name]
-    stdout, _, _ = vm_util.IssueCommand(show_cmd, suppress_warning=True)
-    try:
-      json.loads(stdout)
-    except ValueError:
-      return False
-    return True
-
-  @vm_util.Retry()
-  def _PostCreate(self):
-    """Get the disk's name."""
-    show_cmd = [AZURE_PATH,
-                'vm',
-                'show',
-                '--json',
-                self.vm_name]
-    stdout, _, _ = vm_util.IssueCommand(show_cmd)
+    assert not self.is_image
+
+    vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'storage', 'blob', 'delete',
+         '--container', 'vhds',
+         '--quiet',
+         '%s.vhd' % self.name] +
+        self.storage_account.connection_args)
+
+  def _GetDiskJSON(self):
+    """Get Azure's JSON representation of the disk."""
+    stdout, _, _ = vm_util.IssueCommand(
+        [azure.AZURE_PATH, 'vm', 'disk', 'list',
+         '--json',
+         self.vm_name] + self.resource_group.args)
     response = json.loads(stdout)
 
     # data disks are ordered *alphabetically* by lun, so once we hit
     # lun 10 the ordering doesn't match numerical order. We could
     # compute the right offset for the alphabetical ordering, but it's
     # simpler and more resilient to just search the array, and the
     # size is limited to 15 so it will never take too long.
-    data_disk = (disk for disk in response['DataDisks']
-                 if int(disk.get('logicalUnitNumber', 0)) == self.lun).next()
-    assert ((self.lun == 0 and 'logicalUnitNumber' not in data_disk)
-            or (self.lun == int(data_disk['logicalUnitNumber'])))
-    self.name = data_disk['name']
+    data_disk = (disk for disk in response
+                 if int(disk.get('lun', 0)) == self.lun).next()
```

Comment:
If the generator does not find a value, then a) Azure has changed its behavior or output format and b) we have no sensible way to continue, so raising an exception is the best we can do. Looking at this again, the `assert` is completely redundant with the generator, so I will remove it.